File: Delphi-source-code/MyFinance/MyFinanceAssets/src/uFinanceAssetsBL.pas
unit uFinanceAssetsBL;

interface

uses
  System.SysUtils, System.Classes;

const
  SQL_ASSET_DEPRECIATION_REPORT =
    'SELECT a.AssetId, a.AssetName, a.Value AS OriginalValue, ' +
    'a.DepreciatedValue AS CurrentValue, ' +
    '(a.Value - a.DepreciatedValue) / a.UsefulLifeYears AS AnnualDepreciation, ' +
    'a.UsefulLifeYears - DATEDIFF(YEAR, a.PurchaseDate, GETDATE()) AS YearsRemaining ' +
    'FROM Assets a ORDER BY a.AssetName';

  SQL_ASSET_VALUE_BY_CATEGORY =
    'SELECT a.Category, SUM(a.Value) AS TotalOriginal, ' +
    'SUM(a.DepreciatedValue) AS TotalCurrent, COUNT(*) AS AssetCount ' +
    'FROM Assets a GROUP BY a.Category ORDER BY TotalOriginal DESC';

  SQL_ASSET_MAINTENANCE_SCHEDULE =
    'SELECT m.AssetId, a.AssetName, m.LastMaintenanceDate, m.NextMaintenanceDate, m.EstimatedCost ' +
    'FROM AssetMaintenance m INNER JOIN Assets a ON m.AssetId = a.AssetId ' +
    'ORDER BY m.NextMaintenanceDate';

type
  TDepreciationRecord = record
    AssetId: Integer;
    AssetName: string;
    OriginalValue: Currency;
    CurrentValue: Currency;
    AnnualDepreciation: Currency;
    YearsRemaining: Integer;
  end;

  TAssetCategoryRecord = record
    Category: string;
    TotalOriginal: Currency;
    TotalCurrent: Currency;
    AssetCount: Integer;
  end;

  TMaintenanceRecord = record
    AssetId: Integer;
    AssetName: string;
    LastMaintenance: TDateTime;
    NextMaintenance: TDateTime;
    Cost: Currency;
  end;

  TFinanceAssetsBL = class
  public
    function GetDepreciationReport: TArray<TDepreciationRecord>;
    function GetAssetValueByCategory: TArray<TAssetCategoryRecord>;
    function GetMaintenanceSchedule: TArray<TMaintenanceRecord>;
  end;

implementation

{ TFinanceAssetsBL }

function TFinanceAssetsBL.GetDepreciationReport: TArray<TDepreciationRecord>;
begin
  SetLength(Result, 8);

  Result[0].AssetId := 1;
  Result[0].AssetName := 'Commercial Oven';
  Result[0].OriginalValue := 15000;
  Result[0].CurrentValue := 12000;
  Result[0].AnnualDepreciation := 1500;
  Result[0].YearsRemaining := 6;

  Result[1].AssetId := 2;
  Result[1].AssetName := 'Walk-in Cooler';
  Result[1].OriginalValue := 8000;
  Result[1].CurrentValue := 6500;
  Result[1].AnnualDepreciation := 750;
  Result[1].YearsRemaining := 8;

  Result[2].AssetId := 3;
  Result[2].AssetName := 'POS System';
  Result[2].OriginalValue := 3500;
  Result[2].CurrentValue := 2800;
  Result[2].AnnualDepreciation := 700;
  Result[2].YearsRemaining := 3;

  Result[3].AssetId := 4;
  Result[3].AssetName := 'Dining Furniture Set';
  Result[3].OriginalValue := 12000;
  Result[3].CurrentValue := 7200;
  Result[3].AnnualDepreciation := 1200;
  Result[3].YearsRemaining := 4;

  Result[4].AssetId := 5;
  Result[4].AssetName := 'Delivery Van';
  Result[4].OriginalValue := 25000;
  Result[4].CurrentValue := 21000;
  Result[4].AnnualDepreciation := 3333;
  Result[4].YearsRemaining := 5;

  Result[5].AssetId := 6;
  Result[5].AssetName := 'Industrial Dishwasher';
  Result[5].OriginalValue := 5500;
  Result[5].CurrentValue := 4200;
  Result[5].AnnualDepreciation := 650;
  Result[5].YearsRemaining := 6;

  Result[6].AssetId := 7;
  Result[6].AssetName := 'Security Camera System';
  Result[6].OriginalValue := 2800;
  Result[6].CurrentValue := 2100;
  Result[6].AnnualDepreciation := 560;
  Result[6].YearsRemaining := 2;

  Result[7].AssetId := 8;
  Result[7].AssetName := 'Bar Equipment';
  Result[7].OriginalValue := 6000;
  Result[7].CurrentValue := 4000;
  Result[7].AnnualDepreciation := 857;
  Result[7].YearsRemaining := 3;
end;

function TFinanceAssetsBL.GetAssetValueByCategory: TArray<TAssetCategoryRecord>;
begin
  SetLength(Result, 4);

  Result[0].Category := 'Kitchen Equipment';
  Result[0].TotalOriginal := 28500;
  Result[0].TotalCurrent := 22700;
  Result[0].AssetCount := 3;

  Result[1].Category := 'Furniture';
  Result[1].TotalOriginal := 12000;
  Result[1].TotalCurrent := 7200;
  Result[1].AssetCount := 1;

  Result[2].Category := 'Vehicle';
  Result[2].TotalOriginal := 25000;
  Result[2].TotalCurrent := 21000;
  Result[2].AssetCount := 1;

  Result[3].Category := 'Technology';
  Result[3].TotalOriginal := 6300;
  Result[3].TotalCurrent := 4900;
  Result[3].AssetCount := 2;
end;

function TFinanceAssetsBL.GetMaintenanceSchedule: TArray<TMaintenanceRecord>;
begin
  SetLength(Result, 6);

  Result[0].AssetId := 1;
  Result[0].AssetName := 'Commercial Oven';
  Result[0].LastMaintenance := EncodeDate(2026, 4, 15);
  Result[0].NextMaintenance := EncodeDate(2026, 10, 15);
  Result[0].Cost := 450;

  Result[1].AssetId := 2;
  Result[1].AssetName := 'Walk-in Cooler';
  Result[1].LastMaintenance := EncodeDate(2026, 3, 20);
  Result[1].NextMaintenance := EncodeDate(2026, 9, 20);
  Result[1].Cost := 350;

  Result[2].AssetId := 5;
  Result[2].AssetName := 'Delivery Van';
  Result[2].LastMaintenance := EncodeDate(2026, 5, 10);
  Result[2].NextMaintenance := EncodeDate(2026, 8, 10);
  Result[2].Cost := 280;

  Result[3].AssetId := 6;
  Result[3].AssetName := 'Industrial Dishwasher';
  Result[3].LastMaintenance := EncodeDate(2026, 2, 28);
  Result[3].NextMaintenance := EncodeDate(2026, 8, 28);
  Result[3].Cost := 200;

  Result[4].AssetId := 8;
  Result[4].AssetName := 'Bar Equipment';
  Result[4].LastMaintenance := EncodeDate(2026, 1, 15);
  Result[4].NextMaintenance := EncodeDate(2026, 7, 15);
  Result[4].Cost := 500;

  Result[5].AssetId := 7;
  Result[5].AssetName := 'Security Camera System';
  Result[5].LastMaintenance := EncodeDate(2026, 6, 1);
  Result[5].NextMaintenance := EncodeDate(2026, 12, 1);
  Result[5].Cost := 150;
end;

end.

File: MyFinanceAssets.Core/Application/Services/MyFinanceAssets/uFinanceAssetsBL.cs
using Microsoft.Extensions.Logging;
using MyFinanceAssets.Core.Application.DTOs.Core;

namespace MyFinanceAssets.Core.Application.Services.MyFinanceAssets;

public class uFinanceAssetsBL : IuFinanceAssetsBL
{
    private readonly ILogger<uFinanceAssetsBL> _logger;

    public uFinanceAssetsBL(ILogger<uFinanceAssetsBL> logger)
    {
        _logger = logger;
    }

    public List<TDepreciationRecord> GetDepreciationReport()
    {
        _logger.LogInformation("Generating asset depreciation report (sync).");
        return BuildDepreciationData();
    }

    public async Task<TDepreciationRecord[]> GetDepreciationReportAsync(CancellationToken ct = default)
    {
        _logger.LogInformation("Generating asset depreciation report.");
        await Task.CompletedTask;
        return BuildDepreciationData().ToArray();
    }

    public List<TAssetCategoryRecord> GetAssetValueByCategory()
    {
        _logger.LogInformation("Generating asset value by category report (sync).");
        return BuildCategoryData();
    }

    public async Task<TAssetCategoryRecord[]> GetAssetValueByCategoryAsync(CancellationToken ct = default)
    {
        _logger.LogInformation("Generating asset value by category report.");
        await Task.CompletedTask;
        return BuildCategoryData().ToArray();
    }

    public List<TMaintenanceRecord> GetMaintenanceSchedule()
    {
        _logger.LogInformation("Generating asset maintenance schedule (sync).");
        return BuildMaintenanceData();
    }

    public async Task<TMaintenanceRecord[]> GetMaintenanceScheduleAsync(CancellationToken ct = default)
    {
        _logger.LogInformation("Generating asset maintenance schedule.");
        await Task.CompletedTask;
        return BuildMaintenanceData().ToArray();
    }

    private static List<TDepreciationRecord> BuildDepreciationData()
    {
        return new List<TDepreciationRecord>
        {
            new TDepreciationRecord
            {
                AssetId = 1,
                AssetName = "Commercial Oven",
                OriginalValue = 15000m,
                CurrentValue = 12000m,
                AnnualDepreciation = 1500m,
                YearsRemaining = 6
            },
            new TDepreciationRecord
            {
                AssetId = 2,
                AssetName = "Walk-in Cooler",
                OriginalValue = 8000m,
                CurrentValue = 6500m,
                AnnualDepreciation = 750m,
                YearsRemaining = 8
            },
            new TDepreciationRecord
            {
                AssetId = 3,
                AssetName = "POS System",
                OriginalValue = 3500m,
                CurrentValue = 2800m,
                AnnualDepreciation = 700m,
                YearsRemaining = 3
            },
            new TDepreciationRecord
            {
                AssetId = 4,
                AssetName = "Dining Furniture Set",
                OriginalValue = 12000m,
                CurrentValue = 7200m,
                AnnualDepreciation = 1200m,
                YearsRemaining = 4
            },
            new TDepreciationRecord
            {
                AssetId = 5,
                AssetName = "Delivery Van",
                OriginalValue = 25000m,
                CurrentValue = 21000m,
                AnnualDepreciation = 3333m,
                YearsRemaining = 5
            },
            new TDepreciationRecord
            {
                AssetId = 6,
                AssetName = "Industrial Dishwasher",
                OriginalValue = 5500m,
                CurrentValue = 4200m,
                AnnualDepreciation = 650m,
                YearsRemaining = 6
            },
            new TDepreciationRecord
            {
                AssetId = 7,
                AssetName = "Security Camera System",
                OriginalValue = 2800m,
                CurrentValue = 2100m,
                AnnualDepreciation = 560m,
                YearsRemaining = 2
            },
            new TDepreciationRecord
            {
                AssetId = 8,
                AssetName = "Bar Equipment",
                OriginalValue = 6000m,
                CurrentValue = 4000m,
                AnnualDepreciation = 857m,
                YearsRemaining = 3
            }
        };
    }

    private static List<TAssetCategoryRecord> BuildCategoryData()
    {
        return new List<TAssetCategoryRecord>
        {
            new TAssetCategoryRecord
            {
                Category = "Kitchen Equipment",
                TotalOriginal = 28500m,
                TotalCurrent = 22700m,
                AssetCount = 3
            },
            new TAssetCategoryRecord
            {
                Category = "Furniture",
                TotalOriginal = 12000m,
                TotalCurrent = 7200m,
                AssetCount = 1
            },
            new TAssetCategoryRecord
            {
                Category = "Vehicle",
                TotalOriginal = 25000m,
                TotalCurrent = 21000m,
                AssetCount = 1
            },
            new TAssetCategoryRecord
            {
                Category = "Technology",
                TotalOriginal = 6300m,
                TotalCurrent = 4900m,
                AssetCount = 2
            }
        };
    }

    private static List<TMaintenanceRecord> BuildMaintenanceData()
    {
        return new List<TMaintenanceRecord>
        {
            new TMaintenanceRecord
            {
                AssetId = 1,
                AssetName = "Commercial Oven",
                LastMaintenance = new DateTime(2026, 4, 15, 0, 0, 0, DateTimeKind.Utc),
                NextMaintenance = new DateTime(2026, 10, 15, 0, 0, 0, DateTimeKind.Utc),
                Cost = 450m
            },
            new TMaintenanceRecord
            {
                AssetId = 2,
                AssetName = "Walk-in Cooler",
                LastMaintenance = new DateTime(2026, 3, 20, 0, 0, 0, DateTimeKind.Utc),
                NextMaintenance = new DateTime(2026, 9, 20, 0, 0, 0, DateTimeKind.Utc),
                Cost = 350m
            },
            new TMaintenanceRecord
            {
                AssetId = 5,
                AssetName = "Delivery Van",
                LastMaintenance = new DateTime(2026, 5, 10, 0, 0, 0, DateTimeKind.Utc),
                NextMaintenance = new DateTime(2026, 8, 10, 0, 0, 0, DateTimeKind.Utc),
                Cost = 280m
            },
            new TMaintenanceRecord
            {
                AssetId = 6,
                AssetName = "Industrial Dishwasher",
                LastMaintenance = new DateTime(2026, 2, 28, 0, 0, 0, DateTimeKind.Utc),
                NextMaintenance = new DateTime(2026, 8, 28, 0, 0, 0, DateTimeKind.Utc),
                Cost = 200m
            },
            new TMaintenanceRecord
            {
                AssetId = 8,
                AssetName = "Bar Equipment",
                LastMaintenance = new DateTime(2026, 1, 15, 0, 0, 0, DateTimeKind.Utc),
                NextMaintenance = new DateTime(2026, 7, 15, 0, 0, 0, DateTimeKind.Utc),
                Cost = 500m
            },
            new TMaintenanceRecord
            {
                AssetId = 7,
                AssetName = "Security Camera System",
                LastMaintenance = new DateTime(2026, 6, 1, 0, 0, 0, DateTimeKind.Utc),
                NextMaintenance = new DateTime(2026, 12, 1, 0, 0, 0, DateTimeKind.Utc),
                Cost = 150m
            }
        };
    }
}
